File: frontend/app/utils/workdays.ts
/**
 * 工作日計算工具（前端鏡像後端 internal/utils/workdays.go 的語意）。
 *
 * 統一規則：
 *  - 工期 N 個工作日採「含起始日」語意，結束日 = addWorkingDays(start, N - 1)。
 *  - start 會先正規化到工作日（ensureWorkingDay）。
 *  - 串聯下一階段起點 = nextWorkingDay(prevEnd)。
 *  - 「工作日」= 非週末且非國定假日；補班日例外仍算工作日。
 *  - calendar 模式退化為日曆天，未來可由帳號設定切換。
 *
 * 這些是純函式、不依賴任何 store；元件請透過 useWorkdays() 取得已綁定假日資料的版本。
 */
import { parseISO, format, addDays, isValid } from 'date-fns'

export type DayCalcMode = 'working' | 'calendar'

export interface DayCalcConfig {
  mode: DayCalcMode
  /** 視為休假的星期（0=週日 … 6=週六），預設 [0, 6] */
  weekends: number[]
  /** 國定假日（休假）日期集合，格式 "yyyy-MM-dd" */
  holidays: Set<string>
  /** 補班日（即使落在週末仍須上班）日期集合，格式 "yyyy-MM-dd" */
  makeupWorkdays: Set<string>
}

export const DEFAULT_WEEKENDS: number[] = [0, 6]

/** 全域預設設定：工作日模式、週末為六日、無假日資料（需由 useHolidays 載入）。 */
export function defaultDayCalcConfig(): DayCalcConfig {
  return {
    mode: 'working',
    weekends: [...DEFAULT_WEEKENDS],
    holidays: new Set<string>(),
    makeupWorkdays: new Set<string>()
  }
}

/** 正規化為當天零點，避免時分與時區干擾日期判斷。 */
function toDay(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth(), d.getDate())
}

function toKey(d: Date): string {
  return format(d, 'yyyy-MM-dd')
}

export function isWorkingDay(date: Date, cfg: DayCalcConfig): boolean {
  if (cfg.mode === 'calendar') return true
  const d = toDay(date)
  const key = toKey(d)
  if (cfg.makeupWorkdays.has(key)) return true
  if (cfg.holidays.has(key)) return false
  return !cfg.weekends.includes(d.getDay())
}

/** 若 date 不是工作日則順延到下一個工作日；否則回傳正規化後的當天。 */
export function ensureWorkingDay(date: Date, cfg: DayCalcConfig): Date {
  let d = toDay(date)
  if (cfg.mode === 'calendar') return d
  while (!isWorkingDay(d, cfg)) d = addDays(d, 1)
  return d
}

/** 嚴格晚於 date 的第一個工作日。 */
export function nextWorkingDay(date: Date, cfg: DayCalcConfig): Date {
  let d = addDays(toDay(date), 1)
  if (cfg.mode === 'calendar') return d
  while (!isWorkingDay(d, cfg)) d = addDays(d, 1)
  return d
}

/**
 * 從 start 起算往後加 n 個工作日。
 * start 會先正規化到工作日；含起始日語意（工期 N 的結束日 = addWorkingDays(start, N - 1)）。
 * n 應為非負；calendar 模式退化為日曆天。
 */
export function addWorkingDays(start: Date, n: number, cfg: DayCalcConfig): Date {
  if (cfg.mode === 'calendar') return addDays(toDay(start), n)
  let d = ensureWorkingDay(start, cfg)
  let remaining = n
  while (remaining > 0) {
    d = addDays(d, 1)
    if (isWorkingDay(d, cfg)) remaining--
  }
  return d
}

/**
 * a 到 b 之間的工作日數（不含 a 當天、含 b 當天）；b 早於 a 回傳負值。
 * 用於「距離截止還有幾個工作日」的倒數。calendar 模式等同日曆天差。
 */
export function workingDaysBetween(a: Date, b: Date, cfg: DayCalcConfig): number {
  let da = toDay(a)
  let db = toDay(b)
  if (da.getTime() === db.getTime()) return 0
  let sign = 1
  if (db.getTime() < da.getTime()) {
    const tmp = da
    da = db
    db = tmp
    sign = -1
  }
  let count = 0
  let d = da
  while (d.getTime() < db.getTime()) {
    d = addDays(d, 1)
    if (isWorkingDay(d, cfg)) count++
  }
  return sign * count
}

/**
 * 由開始日字串與工期（工作日數）計算結束日字串。
 * 統一語意：含起始日，start 先正規化到工作日。回傳 "yyyy-MM-dd"；無效輸入回傳 ''。
 */
export function calculateEndDate(startDate: string, durationDays: number, cfg: DayCalcConfig): string {
  if (!startDate || durationDays < 1) return ''
  const parsed = parseISO(startDate)
  if (!isValid(parsed)) return ''
  const start = ensureWorkingDay(parsed, cfg)
  return format(addWorkingDays(start, durationDays - 1, cfg), 'yyyy-MM-dd')
}

File: backend/internal/utils/workdays.go
package utils

import "time"

// dateLayout 是工作日計算統一使用的日期鍵格式。
const dateLayout = "2006-01-02"

// DayCalcMode 控制天數如何換算為實際日期。
type DayCalcMode string

const (
	// DayCalcModeWorking 工作日模式：排除週末與國定假日（補班日除外）。
	DayCalcModeWorking DayCalcMode = "working"
	// DayCalcModeCalendar 日曆日模式：所有日期一律視為可用（行為等同舊邏輯）。
	DayCalcModeCalendar DayCalcMode = "calendar"
)

// DayCalcConfig 描述「天數 → 日期」的換算規則。
//
// 這是整個流程排程（案件階段、流程範本）天數計算的單一真相來源；
// 前端 utils/workdays.ts 鏡像相同語意。
//
// 目前由 DefaultDayCalcConfig() 提供全域預設（工作日 + 國定假日），
// 並預留 per-account 覆寫的 drop-in 介面：未來 resolver 可依帳號改寫
// Mode（切回日曆日）、清空 Holidays（不排除國定假日）或加入自訂假日，
// 而所有呼叫端（AddWorkingDays / NextWorkingDay …）完全不需更動。
type DayCalcConfig struct {
	Mode DayCalcMode
	// Weekends 視為休假的星期（預設週六、週日）。
	Weekends map[time.Weekday]bool
	// Holidays "YYYY-MM-DD" → 國定假日（休假，不算工作日）。
	Holidays map[string]bool
	// MakeupWorkdays "YYYY-MM-DD" → 補班日（台灣特有：即使落在週末仍須上班）。
	MakeupWorkdays map[string]bool
}

// DefaultWeekends 回傳預設週末集合（週六、週日）。
func DefaultWeekends() map[time.Weekday]bool {
	return map[time.Weekday]bool{time.Saturday: true, time.Sunday: true}
}

// DefaultDayCalcConfig 回傳全域預設設定：工作日模式、週末為六日、
// 假日集合為空（需由 resolver 從 national_holidays 載入）。
func DefaultDayCalcConfig() DayCalcConfig {
	return DayCalcConfig{
		Mode:           DayCalcModeWorking,
		Weekends:       DefaultWeekends(),
		Holidays:       map[string]bool{},
		MakeupWorkdays: map[string]bool{},
	}
}

// toDay 取 t 的「當地日曆日」並正規化為 UTC 零點，避免時分秒與時區位移
// 影響日期比較與步進。
func toDay(t time.Time) time.Time {
	y, m, d := t.Date()
	return time.Date(y, m, d, 0, 0, 0, 0, time.UTC)
}

// dateKey 取 t 的 YYYY-MM-DD 日期鍵。
func dateKey(t time.Time) string {
	y, m, d := t.Date()
	return time.Date(y, m, d, 0, 0, 0, 0, time.UTC).Format(dateLayout)
}

// IsWorkingDay 判斷 t 是否為工作日。
//
// 規則優先序：補班日 > 國定假日 > 週末。
// calendar 模式下所有日期皆為工作日。
func (c DayCalcConfig) IsWorkingDay(t time.Time) bool {
	if c.Mode == DayCalcModeCalendar {
		return true
	}
	key := dateKey(t)
	if c.MakeupWorkdays[key] {
		return true
	}
	if c.Holidays[key] {
		return false
	}
	if c.Weekends[t.Weekday()] {
		return false
	}
	return true
}

// EnsureWorkingDay 將 t 正規化為日期；若不是工作日則順延到下一個工作日。
// calendar 模式僅做日期正規化。
func (c DayCalcConfig) EnsureWorkingDay(t time.Time) time.Time {
	d := toDay(t)
	if c.Mode == DayCalcModeCalendar {
		return d
	}
	for !c.IsWorkingDay(d) {
		d = d.AddDate(0, 0, 1)
	}
	return d
}

// NextWorkingDay 回傳嚴格晚於 t 的第一個工作日。
// 用於串聯排程「下一階段從前一階段結束日之後開始」。
func (c DayCalcConfig) NextWorkingDay(t time.Time) time.Time {
	d := toDay(t).AddDate(0, 0, 1)
	if c.Mode == DayCalcModeCalendar {
		return d
	}
	for !c.IsWorkingDay(d) {
		d = d.AddDate(0, 0, 1)
	}
	return d
}

// AddWorkingDays 從 start 起算往後加 n 個工作日。
//
// start 會先正規化到工作日（EnsureWorkingDay）。採「含起始日」語意：
// 工期 N 個工作日的結束日 = AddWorkingDays(start, N-1)，故 n=0 時回傳正規化後的 start。
// n 應為非負；calendar 模式退化為 AddDate(0, 0, n)。
func (c DayCalcConfig) AddWorkingDays(start time.Time, n int) time.Time {
	if c.Mode == DayCalcModeCalendar {
		return toDay(start).AddDate(0, 0, n)
	}
	d := c.EnsureWorkingDay(start)
	for remaining := n; remaining > 0; {
		d = d.AddDate(0, 0, 1)
		if c.IsWorkingDay(d) {
			remaining--
		}
	}
	return d
}

// WorkingDaysBetween 回傳 a 到 b 之間的工作日數（不含 a 當天、含 b 當天）。
// b 早於 a 時回傳負值。用於「距離截止還有幾個工作日」的倒數計算。
// calendar 模式等同日曆天差。
func (c DayCalcConfig) WorkingDaysBetween(a, b time.Time) int {
	da, db := toDay(a), toDay(b)
	if da.Equal(db) {
		return 0
	}
	sign := 1
	if db.Before(da) {
		da, db = db, da
		sign = -1
	}
	count := 0
	for d := da; d.Before(db); {
		d = d.AddDate(0, 0, 1)
		if c.IsWorkingDay(d) {
			count++
		}
	}
	return sign * count
}
